### PR TITLE
Image prompt template

### PR DESCRIPTION
**Pull request by hinthornw**
Builds on Bagatur's (#13227). See unit test for example usage (below)

```python
def test_chat_tmpl_from_messages_multipart_image() -> None:
    base64_image = "abcd123"
    other_base64_image = "abcd123"
    template = ChatPromptTemplate.from_messages(
        [
            ("system", "You are an AI assistant named {name}."),
            (
                "human",
                [
                    {"type": "text", "text": "What's in this image?"},
                    # OAI supports all these structures today
                    {
                        "type": "image_url",
                        "image_url": "data:image/jpeg;base64,{my_image}",
                    },
                    {
                        "type": "image_url",
                        "image_url": {"url": "data:image/jpeg;base64,{my_image}"},
                    },
                    {"type": "image_url", "image_url": "{my_other_image}"},
                    {
                        "type": "image_url",
                        "image_url": {"url": "{my_other_image}", "detail": "medium"},
                    },
                    {
                        "type": "image_url",
                        "image_url": {"url": "https://www.langchain.com/image.png"},
                    },
                    {
                        "type": "image_url",
                        "image_url": {"url": "data:image/jpeg;base64,foobar"},
                    },
                ],
            ),
        ]
    )
    messages = template.format_messages(
        name="R2D2", my_image=base64_image, my_other_image=other_base64_image
    )
    expected = [
        SystemMessage(content="You are an AI assistant named R2D2."),
        HumanMessage(
            content=[
                {"type": "text", "text": "What's in this image?"},
                {
                    "type": "image_url",
                    "image_url": {"url": f"data:image/jpeg;base64,{base64_image}"},
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/jpeg;base64,{other_base64_image}"
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {"url": f"{other_base64_image}"},
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"{other_base64_image}",
                        "detail": "medium",
                    },
                },
                {
                    "type": "image_url",
                    "image_url": {"url": "https://www.langchain.com/image.png"},
                },
                {
                    "type": "image_url",
                    "image_url": {"url": "data:image/jpeg;base64,foobar"},
                },
            ]
        ),
    ]
    assert messages == expected
```

### CODE REVIEW

**Review by rlancemartin — APPROVED** [review]
Tested - 
https://github.com/langchain-ai/langchain/pull/14263
https://github.com/langchain-ai/langchain-benchmarks/pull/106

File: libs/core/langchain_core/utils/image.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import mimetypes
+
+
+def encode_image(image_path: str) -> str:
+    """Get base64 string from image URI."""
+    with open(image_path, "rb") as image_file:
+        return base64.b64encode(image_file.read()).decode("utf-8")
+
+
+def image_to_data_url(image_path: str) -> str:
```
Comment (hinthornw) [inline libs/core/langchain_core/utils/image.py:11]: We could also make these accessible as an instance of `ImagePromptValue.to_b64_string` etc.

File: libs/core/langchain_core/prompts/image.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """An image prompt template for a multimodal model."""
+
+    template: dict = Field(default_factory=dict)
+    """Template for the prompt."""
+
+    def __init__(self, **kwargs: Any) -> None:
+        if "input_variables" not in kwargs:
+            kwargs["input_variables"] = []
+
+        overlap = set(kwargs["input_variables"]) & set(("url", "path", "detail"))
```
Comment (hinthornw) [inline libs/core/langchain_core/prompts/image.py:19]: These actually aren't required to be blocked out now that I'm using the fstring format to infer again

**Comment by baskaryan** [comment]
this type of prompt wouldn't work anymore i take it?
```python
prompt = ChatPromptTemplate.from_messages([
    ("system", "foo bar"),
    ("human", [
        "{question}",
        {"text": "{second_question}"},
        # image with var name 'first_image' and no preset values
        {"image_url": "{first_image}"}, 
        # image with var name 'second_image' and preset "detail" level
        {"image_url": {"variable_name": "second_image", "detail": "low"}}, 
        # constant image, not a var
        {"type": "image_url", "image_url": {"url": "https://python.langchain.com/assets/images/langchain_stack-f195a2c417953cb5828f76e5f9039c9f.png"}}]), 
])
```

**Comment by hinthornw** [comment]
```
   {"image_url": {"variable_name": "second_image", "detail": "low"}}, 
```
^ this one wouldn't work right now. Is that a preferable format?

**Review by baskaryan — APPROVED** [review]

**Pull request merged**